executor/containerdexecutor/executor.go

```diff
@@ -329,5 +316,82 @@ func (w *containerdExecutor) Exec(ctx context.Context, id string, process execut
 	if err != nil {
 		return errors.WithStack(err)
 	}
-	return taskProcess.Start(ctx)
+
+	if sc, ok := stdin.(*stdinCloser); ok {
+		sc.closer = func() {
+			taskProcess.CloseIO(ctx, containerd.WithStdinCloser)
+		}
+	}
+
+	err = w.runProcess(ctx, taskProcess, process.Resize, nil)
+	return err
+}
+
+func (w *containerdExecutor) runProcess(ctx context.Context, p containerd.Process, resize <-chan executor.WinSize, started func()) error {
+	statusCh, err := p.Wait(context.Background())
+	if err != nil {
+		return err
+	}
+
+	err = p.Start(ctx)
+	if err != nil {
+		return err
+	}
+
+	if started != nil {
+		started()
+	}
+
+	var cancel func()
+	ctxDone := ctx.Done()
+	for {
+		select {
+		case <-ctxDone:
+			ctxDone = nil
+			var killCtx context.Context
+			killCtx, cancel = context.WithTimeout(context.Background(), 10*time.Second)
+			p.Kill(killCtx, syscall.SIGKILL)
+		case size := <-resize:
+			err := p.Resize(ctx, size.Cols, size.Rows)
+			if err != nil {
+				cancel()
+				return err
+			}
+		case status := <-statusCh:
+			if cancel != nil {
+				cancel()
+			}
+			if status.ExitCode() != 0 {
+				exitErr := &executor.ExitError{
+					ExitCode: status.ExitCode(),
+					Err:      status.Error(),
+				}
+				if status.ExitCode() == containerd.UnknownExitStatus && status.Error() != nil {
+					exitErr.Err = errors.Wrap(status.Error(), "failure waiting for process")
+				}
+				select {
+				case <-ctx.Done():
+					exitErr.Err = errors.Wrap(ctx.Err(), exitErr.Error())
+				default:
+				}
+				return exitErr
+			}
+			return nil
+		}
+	}
+}
+
+type stdinCloser struct {
+	stdin  io.ReadCloser
+	closer func()
+}
+
+func (s *stdinCloser) Read(p []byte) (int, error) {
+	n, err := s.stdin.Read(p)
+	if err == io.EOF {
+		if s.closer != nil {
```

Comment:
I'm not sure if this is safe. Bytes are read and then written to the fifo. Only after the write is complete can we close the fifo. If we close on read there should be a race between writing the last buffer and closing.

Also, not sure about that `CloseIO` call. Afaik it signals the shim and closes stdin there. Not even sure what stdin is in shim and how it is safe to synchronize it (eg. if it is the other side of fifo). I think much more safer would be to just close the fifo on client side after writing the last message.

Reply:
Actually, there is a close in https://github.com/moby/buildkit/blob/master/vendor/github.com/containerd/containerd/cio/io_unix.go#L70 so not sure. So I think `CloseIO` just needs to be called after fifo has been open and task created. This is based on looking at the Docker code. But quite confusing though. If this is true the we shouldn't wait for `EOF`, just close early and we can test that calling `CloseIO` doesn't really drop stdin before fifo gets closed as well.

Reply:
I can test early CloseIO.  I lifted this code from containerd, so I was hoping their logic is correct :) 
https://github.com/containerd/containerd/blob/master/cmd/ctr/commands/tasks/exec.go#L131-L133
https://github.com/containerd/containerd/blob/master/cmd/ctr/commands/tasks/exec.go#L181-L194

Reply:
btw, I did test closing the fifo on client side, it had no impact. The signal to the server side via CloseIO was the only think that allowed the container to exit on EOF on Stdin.  I did not test the ordering of CloseIO though.

Reply:
I see. But can't think of any case how it would be correct that this call happens between reading EOF from a go reader and writing last bytes to a fifo. That time is only relevant in client-side and not synchronized with the server-side at all.

Reply:
I think it is working without the stdin wrapper now.  I also found there was an issue on pid1 when stdin closed.  I have added another test to verify closing stdin on pid1 will cause the container to exit normally.

First tried calling CloseIO after the container.NewTask and task.Exec (after fifo created) .. this worked once, then retries failed, so was racey.
Next tried calling CloseIO after task.Wait, which worked more frequently, but also failed repeatedly, so still racey.
Finally tried calling CloseIO after task.Start, which seems to work consistently.

Anyway, it appears you are right, we can call CloseIO any time after the container is running.  Not sure why the `ctr` code is written that way.

Reply:
@crosbymichael fyi